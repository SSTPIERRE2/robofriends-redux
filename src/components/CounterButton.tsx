import React, { Component } from 'react';

interface State {
    count: number;
}

class CounterButton extends Component<Record<string, unknown>, State> {
    constructor(props: null) {
        super(props);
        this.state = { count: 1 };
    }

    shouldComponentUpdate(nextProps: null, nextState: State) {
        if (this.state.count !== nextState.count) {
            return true;
        }
        return false;
    }

    render() {
        return (
            <button
                onClick={() =>
                    this.setState((state) => ({ count: state.count + 1 }))
                }
            >
                Count: {this.state.count}
            </button>
        );
    }
}

export default CounterButton;
