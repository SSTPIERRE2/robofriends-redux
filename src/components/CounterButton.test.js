import { shallow, mount } from "enzyme";
import CounterButton from "./CounterButton";
import React from "react";
import sinon from "sinon";

describe("CounterButton", () => {
  const sandbox = sinon.createSandbox();

  afterEach(function () {
    sandbox.restore();
  });

  it("renders the CounterButton UI", () => {
    expect(shallow(<CounterButton />)).toMatchSnapshot();
  });

  it("has an initial count of 1", () => {
    const wrapper = mount(<CounterButton />);
    expect(wrapper.state("count")).toBe(1);
  });

  it("increases count by 1 when clicked", () => {
    sandbox.spy(CounterButton.prototype, "shouldComponentUpdate");
    const wrapper = mount(<CounterButton />);
    wrapper.find("button").simulate("click");
    expect(wrapper.state("count")).toBe(2);
    expect(CounterButton.prototype.shouldComponentUpdate).toHaveProperty(
      "callCount",
      1
    );
    expect(CounterButton.prototype.shouldComponentUpdate.returned(true)).toBe(
      true
    );
  });

  it("does not update if nextState's count is the same as the current state's count", () => {
    sandbox.spy(CounterButton.prototype, "shouldComponentUpdate");
    const wrapper = mount(<CounterButton />);
    wrapper.setState({ count: 1 });
    expect(CounterButton.prototype.shouldComponentUpdate).toHaveProperty(
      "callCount",
      1
    );
    expect(CounterButton.prototype.shouldComponentUpdate.returned(false)).toBe(
      true
    );
  });
});
