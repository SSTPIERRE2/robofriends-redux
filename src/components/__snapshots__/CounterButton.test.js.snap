// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterButton renders the CounterButton UI 1`] = `ShallowWrapper {}`;
